flasc/visualization.py

```diff
@@ -637,14 +639,137 @@ def put_label(i):
         raise ValueError("Need one of near_i, near_x, near_y")
 
 
+def plot_binned_mean_and_ci(
+    x,
+    y,
+    color="b",
+    label="_nolegend_",
+    x_edges=None,
+    ax=None,
+    show_scatter=True,
+    show_bin_points=True,
+    show_confidence=True,
+    alpha_scatter=0.1,
+    confidence_level = 0.95,
+):
+    """
+    Plot data to a single axis.  Method
+    has options to include scatter of underlying data, specifiying
+    bin edges, and plotting confidence interval.
 
+    Args:
+        x (np.array): abscissa data.
+        y (np.array): ordinate data.
+        color (str, optional): line color.
+            Defaults to 'b'.
+        label (str, optional): line label used in legend.
+            Defaults to '_nolegend_'.
+        x_edges (np.array, optional): bin edges in x data
+            Defaults to None.
+        ax (:py:class:`matplotlib.pyplot.axes`, optional):
+            axes handle for plotting. Defaults to None.
+        show_scatter (bool, optional): flag to control scatter plot.
+            Defaults to True.
+        show_bin_points (bool, optional): flag to control plot of bins.
+            Defaults to True.
+        show_confidence (bool, optional): flag to control plot of
+            confidence interval. Defaults to True.
+        alpha_scatter (float, optional): Alpha for scatter
+            plot. Defaults to 0.5.
+        confidenceLevel (float, optional): Confidence level for
+            confidence interval. Defaults to 0.95.
 
+    """
 
+    # Check the length of x equals length of y
+    if len(x) != len(y):
+        raise ValueError("x and y must be the same length")
     
+    # Check that x is not empty
+    if len(x) == 0:
+        raise ValueError("x is empty")
     
 
-    
+    # Declare ax if not provided
+    if ax is None:
+        _, ax = plt.subplots()
 
+    # Put points ino dataframe
+    df = pd.DataFrame({"x": x, "y": y})
 
+    # If x_edges not provided, use 50 bins over range of x
+    if x_edges is None:
+        x_edges = np.linspace(df["x"].min()*.98, df["x"].max()*1.02, 50)
 
-    
+    # Define x_labels as bin centers
+    x_labels = (x_edges[1:] + x_edges[:-1]) / 2.0
+
+    # Bin data
+    df["x_bin"] = pd.cut(df["x"], x_edges, labels=x_labels)
+
+    # Get aggregate statistics
+    df_agg = df.groupby("x_bin").agg(
+        {"y": ["count", "std", "min", "max", "mean", st.sem]}
+    )
+    # Flatten column names
+    df_agg.columns = ["_".join(c) for c in df_agg.columns]
+
+    # Reset the index
+    df_agg = df_agg.reset_index()
+
+    # Delete rows with no data
+    df_agg = df_agg[df_agg["y_count"] > 0]
+
+    # Add the confidence interval of the mean to df_agg
+    df_agg["y_ci_lower"], df_agg["y_ci_upper"] = st.t.interval(
+        confidence_level, 
+        df_agg["y_count"]-1,
+        loc=df_agg["y_mean"],
+        scale=df_agg["y_sem"]
+    )
```

Comment:
@paulf81 I think what you have, using the t-distribution, is good (it's a little more correct than the z-score approach, which assumes the data set is large enough that the means can be treated as normal---the t-distribution doesn't make that assumption, and approaches the z-score method as the number of samples becomes large). If rearranged the code a little in my commits, but have not changed the method.